src/java/org/apache/cassandra/service/reads/DataResolver.java

```diff
@@ -135,6 +135,9 @@ private boolean needsReplicaFilteringProtection()
         if (command.rowFilter().isEmpty())
             return false;
 
+        if (command.isTopK())
+            return false;
```

Comment:
This method is called `needsReplicaFilteringProtection`, but it actually is a mix of whether the query needs and supports replica filtering protection. Maybe we could rename it to `useReplicaFilteringProtection`.

I understand that top-k queries with CL_ONE do need replica filtering protection, but we aren't able to provide it, is that right?

If a top-k query with CL>ONE gets a row from one replica and nothing from the other, we cannot know if that row has been changed in the other node. That can lead to the resurrection of deleted/stale rows. Also, read repair might send the stale row to the silent replica. Am I missing something in top-k queries that prevents this situation? I haven't read through all the patch yet so I might be missing something.

I guess a possible approach would be just forbidding top-k queries with CL>ONE. 

I understand that ANN in particular, differently to general top-k queries, only has to produce approximate results. So I guess we could use RFP for ANN queries and, if the silent replica gives us something that is not very far away from the index result, just let it pass.

A third approach could be simply failing the queries if RFP actually finds a silent replica. In theory, this shouldn't be a very common situation. Or, instead of failing, it could trigger RR and query the silent replica again. This later approach, if viable, would be quite involved so we could do it in a separate patch.

Reply:
Here is a dtest showing how a vector search with CL>ONE can wrongly return a row that has been deleted:
```java
try (Cluster cluster = init(Cluster.build(2).withConfig(config -> config.with(GOSSIP, NETWORK)).start()))
{
    cluster.schemaChange(withKeyspace("CREATE TABLE %s.t (k int PRIMARY KEY, v vector<float, 1>)"));
    cluster.schemaChange(withKeyspace("CREATE CUSTOM INDEX ON %s.t(v) USING 'StorageAttachedIndex'"));
    SAIUtil.waitForIndexQueryable(cluster, KEYSPACE);

    // insert a row in both nodes, and then remove in only one of the nodes
    cluster.coordinator(1).execute(withKeyspace("INSERT INTO %s.t (k, v) VALUES (1, [1])"), ConsistencyLevel.QUORUM);
    cluster.get(1).executeInternal(withKeyspace("DELETE FROM %s.t WHERE k=1"));

    String query = withKeyspace("SELECT * FROM %s.t ORDER BY v ANN OF [1] LIMIT 10");
    Object[][] rows = cluster.coordinator(1).execute(query, ConsistencyLevel.QUORUM);
    assertThat(rows).isEmpty(); // wrongly returns the deleted row
}
```
This other

Reply:
And, since RFP is disabled for all the query, if we restrict any other column we can also return stale rows that don't satisfy that filter:
```java
try (Cluster cluster = init(Cluster.build(2).start()))
{
    cluster.schemaChange(withKeyspace("CREATE TABLE %s.t (k int PRIMARY KEY, v vector<float, 1>, c text)"));
    cluster.schemaChange(withKeyspace("CREATE CUSTOM INDEX ON %s.t(v) USING 'StorageAttachedIndex'"));
    cluster.schemaChange(withKeyspace("CREATE CUSTOM INDEX ON %s.t(c) USING 'StorageAttachedIndex'"));

    // insert a row in both nodes, and then update it in only one of the nodes
    cluster.coordinator(1).execute(withKeyspace("INSERT INTO %s.t (k, v, c) VALUES (1, [1], 'old')"), ConsistencyLevel.QUORUM);
    cluster.get(1).executeInternal(withKeyspace("UPDATE %s.t SET c='new' WHERE k=1"));

    String query = withKeyspace("SELECT * FROM %s.t WHERE c='old' ORDER BY v ANN OF [1] LIMIT 1");
    Object[][] rows = cluster.coordinator(1).execute(query, ConsistencyLevel.QUORUM);
    assertThat(rows).isEmpty(); // wrongly returns the stale row, (1, old, [1.0]), which doesn't satisfy the query
}
```

Reply:
Thanks for this analysis.

I feel that, for this initial implementation, we should restrict the queries to CL=ONE. Should we allow CL=LOCAL_ONE as well? I'll raise a separate ticket to see how we can improve this.

I'll also add a separate dtest to cover this situation as you have shown above.

Reply:
Great. Now that we don't support queries with CL > ONE the `DataResolver` won't be used in top-k queries. I think we can remove all the changes on this class disabling SRP, RFP and RR if the command is top-k.